src/artifacts.js

```diff
@@ -183,8 +195,70 @@ api.declare({
   // Construct details for different storage types
   var isPublic = /^public\//.test(name);
   var details  = {};
+  let present = false;
+  let uploadId;
   switch (storageType) {
+    case 'blob':
+      // Generate the details that we'll be using.
+      details.contentLength = input.contentLength;
+      details.contentSha256 = input.contentSha256;
+      if (input.transferLength) {
+        details.transferLength = input.transferLength;
+      }
+      if (input.transferSha256) {
+        details.transferSha256 = input.transferSha256;
+      }
+      // We want to ensure, for idempotency reasons, that any following
+      // requests to createArtifact() use the same set of part information.
+      // Instead of storing the entire parts list in the entity, we'll instead
+      // store a hash of the JSON serialized version of the list.
+      //
+      // Note that this means that for loaded entities we'll need to use the
+      // details.partsHash field to check whether we're multipart or not
+      // instead of the details.parts list
+      if (input.parts) {
+        let partsHash = crypto.createHash('sha256');
+        for (let part of input.parts) {
+          partsHash.update(`${part.sha256}${part.size}`);
```

Comment:
May I suggest `${part.sha256}:${part.size}\n`, I know it's nit... because json schema ensures you can't provide these of arbitrary lengths... But it's a nice extra safety feature..

Reply:
we could do that, but this string is never shown anywhere and is just used to generate a sha256 sum.  First, I'd rather not use `\n` because that maps to different characters on different systems.  Second, we're never showing this string anywhere, so the actual format is completely unimportant, so having the simplest format is fine.

I'll change the string to `${part.sha256}:${part.size}_` so that in the off chance we did want to display the string, we can without adding complexity to the string generation.

Reply:
`\n` is NOT platform dependent... `\n` is LF (line feed), `\r` is CR (carriage return).
https://en.wikipedia.org/wiki/Newline

`\n` is not newline (yes, newline is platform dependent), but you'll rarely see anyone use anything but LF for internal strings...